homeassistant/components/homewizard_energy/config_flow.py

```diff
@@ -0,0 +1,182 @@
+            return self.async_abort(reason="api_not_enabled")
+
+        device_info = await self._async_try_connect_and_fetch(discovery_info["host"])
+
+        # Pass parameters
+        self.entry_info = {
+            CONF_IP_ADDRESS: discovery_info["host"],
+            CONF_PORT: discovery_info["port"],
+            CONF_PRODUCT_TYPE: device_info[CONF_PRODUCT_TYPE],
+            CONF_PRODUCT_NAME: device_info[CONF_PRODUCT_NAME],
+            CONF_SERIAL: device_info[CONF_SERIAL],
+        }
+        return await self.async_step_discovery_confirm()
+
+    async def async_step_discovery_confirm(
+        self, entry_info: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Confirm discovery."""
+        if entry_info is not None:
+            return await self._async_create_entry(self.entry_info)
+
+        self._set_confirm_only()
+
+        placeholders = {
+            CONF_PRODUCT_TYPE: self.context["entry_info"][CONF_PRODUCT_TYPE],
+            CONF_SERIAL: self.context["entry_info"][CONF_SERIAL],
+            CONF_IP_ADDRESS: self.context["entry_info"][CONF_IP_ADDRESS],
+        }
+        return self.async_show_form(
+            step_id="discovery_confirm", description_placeholders=placeholders
+        )
+
+    async def _async_try_connect_and_fetch(self, ip_address: str) -> dict[str, Any]:
+        """Try to connect."""
+        # Make connection with device
+        # This is to test the connection and to get info for unique_id
+        energy_api = aiohwenergy.HomeWizardEnergy(ip_address)
+
+        initialized = False
+        try:
+            with async_timeout.timeout(10):
+                await energy_api.initialize()
+                if energy_api.device is not None:
+                    initialized = True
+
+        except aiohwenergy.DisabledError as ex:
+            _LOGGER.error("API disabled, API must be enabled in the app")
+            raise AbortFlow("api_not_enabled") from ex
+
+        except Exception as ex:  # pylint: disable=broad-except
+            _LOGGER.error(
+                f"Error connecting with Energy Device at {ip_address}",
+            )
+            raise AbortFlow("unknown_error") from ex
+
+        finally:
+            await energy_api.close()
+
+        if not initialized:
+            _LOGGER.error("Initialization failed")
+            raise AbortFlow("unknown_error")
+
+        # Validate metadata
+        if energy_api.device.api_version != "v1":
+            raise AbortFlow("unsupported_api_version")
+
+        if energy_api.device.product_type not in SUPPORTED_DEVICES:
+            _LOGGER.error(
+                "Device (%s) not supported by integration",
+                energy_api.device.product_type,
+            )
+            raise AbortFlow("device_not_supported")
+
+        return {
+            CONF_PRODUCT_NAME: energy_api.device.product_name,
+            CONF_PRODUCT_TYPE: energy_api.device.product_type,
+            CONF_SERIAL: energy_api.device.serial,
```

Comment:
Why store the serial in the config entry? You're going to fetch it each time in the coordinator and it's also stored as unique ID.

Reply:
https://github.com/home-assistant/core/pull/55812/commits/fc1c4b7c74592b33f940f24046040973fc9cefed